### PR TITLE
feat: add coursier support

### PR DESCRIPTION
**Pull request by sihil**
as it's increasingly common and the bundled `coursierDependencyGraph`
means there is no need to pollute projects with `sbt-dependency-graph`
(it's also a lot faster as a bonus)

- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Adds support for the sbt-coursier plugin (removing the sbt-dependency-graph requirement). 

#### Where should the reviewer start?
There are two fundamental changes:
 - the command executed in SBT
 - the parsing code that deals with the format differences between the two dependency trees

The new fixtures are a good place to start to understand the differences in format and how it is configured. After that I'd look at the changes in `index.js` where we detect coursier and then finally the parsing code.

Note that because SBT is somewhat slow we try and guess whether coursier is configured by looking for `sbt-coursier` in any files matching `project/*.sbt` or `project/project/*.sbt` (the meta project).

#### How should this be manually tested?
Switching any SBT project to sbt-coursier and checking the tool still works (faster 🤞) would be a great starting point. It goes without saying that it should also work on existing projects.

#### Any background context you want to provide?
I'm principally a Scala dev not a Node.JS dev so may well have missed best practices or other neat tricks.

#### Additional questions
I've not actually tested this within the context of the snyk CLI tool, I'm just trusting that the tests are enough. Very happy to do so if someone can point me in the right direction though.

### CODE REVIEW

**Comment by sihil** [comment]
I've had a look at the CI failing on node 4 and can reproduce the error on master so currently don't believe it is anything to do with my changes. I did note that my `npm install` on node 4 has lots of warnings `wanted: {"node":">=6"}` so perhaps a library version needs to be pinned to make node 4 still work.
Also node 4 is EOL for almost a year, so perhaps dropping it is OK too.

**Comment by lili2311** [comment]
@sihil 👋 Thank you for this awesome contribution, to make sure we fully understand the impact and scope of the proposed feature before pushing it in I would love if we can arrange a time to meet and go through this and to review the and we can use the time to also pair on the tests and get those passing.

**Comment by sihil** [comment]
You're welcome @lili2311. I'm happy to meet if it would be helpful to run through. Also happy to fix tests, although suggest that you folks confirm that `master` is passing beforehand. I'll ping you an e-mail.

**Comment by lili2311** [comment]
👋 @sihil This has not been forgotten. Just chatting internally again to arrange for this to be incorporated.  Apologies for the slow turn around

**Comment by sihil** [comment]
Great! Thanks @lili2311 - let me know if there is anything else that you need from me.

**Comment by sihil** [comment]
It's worth noting that coursier will become the default in the next SBT release - although this plugin will not work as it stands as the detection method is too naive.

https://github.com/sbt/sbt/pull/4614

**Comment by sihil** [comment]
Hi @lili2311 - just wanted to check you'd seen my most recent e-mail regarding reviewing this.

**Comment by sihil** [comment]
Also - I found this open issue to support machine readable output which would be safer than parsing the tree: https://github.com/coursier/coursier/issues/575

**Comment by lili2311** [comment]
Please squash commits into 1 or multiple if there are bits that go together better all following the`feat:` `chore:` `test:` in the commit.

**Comment by lili2311** [comment]
Please squash commits into 1 or multiple if there are bits that go together better all following the`feat:` `chore:` `test:` in the commit.

File: lib/index.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 var path = require('path');
+var fs = require('fs');
```
[the remark this answers is not in the document]
Comment (sihil) [reply lib/index.js]: discussed IRL

File: lib/index.js [hunk 2]
```diff
@@ -33,23 +53,62 @@ function inspect(root, targetFile, options) {
...
+  return buffer.indexOf('sbt-coursier') > -1;
+}
+
+function buildArgs(sbtArgs, isCoursierProject) {
+  let args = ['-Dsbt.log.noformat=true']; // force plain output
   if (sbtArgs) {
     args = args.concat(sbtArgs);
   }
-  args.push('dependencyTree');
+  if (isCoursierProject) {
```
[the remark this answers is not in the document]
Comment (sihil) [reply lib/index.js]: discussed IRL

File: lib/index.js [hunk 3]
```diff
@@ -33,23 +53,62 @@ function inspect(root, targetFile, options) {
...
+  return [];
+}
+
+function searchWithFs( filename ) {
+  const buffer = fs.readFileSync(filename);
+  return buffer.indexOf('sbt-coursier') > -1;
+}
+
+function buildArgs(sbtArgs, isCoursierProject) {
+  let args = ['-Dsbt.log.noformat=true']; // force plain output
```
[the remark this answers is not in the document]
Comment (sihil) [reply lib/index.js]: Updated to mention the colour codes.

**Pull request merged**